src/codegen/codegen_compatibility_visitor.cpp

```diff
@@ -0,0 +1,193 @@
+            }
+            if (token == "bbcore_write") {
+                found_bbcore_write = true;
+            }
+        }
+    }
+    if (!found_bbcore_read) {
+        error_message_no_bbcore_read_write
+            << "\"bbcore_read\" function not defined in any VERBATIM block\n";
+    }
+    if (!found_bbcore_write) {
+        error_message_no_bbcore_read_write
+            << "\"bbcore_write\" function not defined in any VERBATIM block\n";
+    }
+    return error_message_no_bbcore_read_write.str();
+}
+
+/**
+ * \details Checks all the ast::AstNodeType that are not handled in NMODL code
+ * generation backend for CoreNEURON and prints related messages. If there is
+ * some kind of incompatibility return false.
+ */
+bool CodegenCompatibilityVisitor::find_unhandled_ast_nodes(Ast* node) {
+    unhandled_ast_nodes = AstLookupVisitor().lookup(node, unhandled_ast_types);
+
+    std::stringstream ss;
+    for (auto it: unhandled_ast_nodes) {
+        auto node_type = it->get_node_type();
+        switch (node_type) {
+        case AstNodeType::SOLVE_BLOCK: {
+            auto it_solve_block = std::dynamic_pointer_cast<ast::SolveBlock>(it);
+            ss << return_error_if_solve_method_is_unhandled(it_solve_block);
+        } break;
+        case AstNodeType::DISCRETE_BLOCK:
+            ss << return_error_with_name<DiscreteBlock>(it);
+            break;
+        case AstNodeType::PARTIAL_BLOCK:
+            ss << return_error_with_name<PartialBlock>(it);
+            break;
+        case AstNodeType::BEFORE_BLOCK:
+            ss << return_error_without_name<BeforeBlock>(it);
+            break;
+        case AstNodeType::AFTER_BLOCK:
+            ss << return_error_without_name<AfterBlock>(it);
+            break;
+        case AstNodeType::MATCH_BLOCK:
+            ss << return_error_without_name<MatchBlock>(it);
+            break;
+        case AstNodeType::CONSTANT_BLOCK:
+            ss << return_error_without_name<ConstantBlock>(it);
+            break;
+        case AstNodeType::CONSTRUCTOR_BLOCK:
+            ss << return_error_without_name<ConstructorBlock>(it);
+            break;
+        case AstNodeType::DESTRUCTOR_BLOCK:
+            ss << return_error_without_name<DestructorBlock>(it);
+            break;
+        case AstNodeType::INDEPENDENT_BLOCK:
+            ss << return_error_without_name<IndependentBlock>(it);
+            break;
+        case AstNodeType::FUNCTION_TABLE_BLOCK:
+            ss << return_error_without_name<FunctionTableBlock>(it);
+            break;
+        case AstNodeType::GLOBAL_VAR: {
+            auto it_global_var = std::dynamic_pointer_cast<ast::GlobalVar>(it);
+            ss << return_error_global_var(node, it_global_var);
+        } break;
+        case AstNodeType::POINTER_VAR: {
+            auto it_pointer_var = std::dynamic_pointer_cast<ast::PointerVar>(it);
+            ss << return_error_pointer(it_pointer_var);
+        } break;
+        case AstNodeType::BBCORE_POINTER_VAR:
```

Comment:
would it not be better to have vectors/sets of AstNodeType the ones that shoudl be printed with name, the ones without and then iterate over those two separately? This big switch case doesn't look very nice in the end and seems a bit too verbose?

Reply:
It's true that it doesn't look very nice but I had a hard time to make it less verbose. I have used templates to call the `return_error` functions to avoid declaring lot's of different functions that do the same thing, so for every different type of `node` there must be a switch case that calls the corresponding function with the correct template argument. What could be done is divide the `AstNodeType`s to those that should be printed with name and without as you suggested and then divide also the large switch case in smaller parts or functions. Would this be enough? I also tried to not overcomplicate this part, so that in the future it might be easier to add more cases if needed in the same place.

Reply:
(discussed offline a possible solution)